src/main/java/com/mcmoddev/communitymod/SubModLoader.java

```diff
@@ -0,0 +1,217 @@
+package com.mcmoddev.communitymod;
+
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.fml.common.discovery.ASMDataTable;
+import net.minecraftforge.fml.common.event.FMLConstructionEvent;
+import net.minecraftforge.registries.IForgeRegistry;
+import net.minecraftforge.registries.IForgeRegistryEntry;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.*;
+
+public enum SubModLoader
+{
+    INSTANCE;
+
+    private static final Logger LOGGER = LogManager.getLogger("SubModLoader");
+
+    private final Map<IForgeRegistryEntry, SubModContainer> regEntryOrigin = new HashMap<>();
+    private final Map<ISubMod, SubModContainer> subMods = new HashMap<>();
+    private final List<SubModContainer> loadedSubMods = new ArrayList<>();
+
+    public void onConstruction(FMLConstructionEvent event)
+    {
+        long startTime = System.currentTimeMillis();
+        Set<SubModContainer> eventSubscriberWarns = new HashSet<>();
+        Set<ASMDataTable.ASMData> dataSet = event.getASMHarvestedData().getAll(SubMod.class.getCanonicalName());
+
+        for (ASMDataTable.ASMData data : dataSet)
+        {
+            String modName = (String) data.getAnnotationInfo().get("name");
+
+            try
+            {
+                if (!data.getAnnotationInfo().containsKey("clientSideOnly") || (boolean) data.getAnnotationInfo().get("clientSideOnly") && event.getSide().isClient())
+                {
+                    ISubMod subMod = (ISubMod) Class.forName(data.getClassName()).newInstance();
+                    String description = (String) data.getAnnotationInfo().getOrDefault("description", "No description is provided for this submod.");
+                    String attribution = (String) data.getAnnotationInfo().getOrDefault("attribution", "No attribution is provided for this submod.");
+                    Boolean requiresMcRestart = (Boolean) data.getAnnotationInfo().getOrDefault("requiresMcRestart", true);
+                    SubModContainer container = new SubModContainer(modName, description, attribution, subMod, requiresMcRestart);
+
+                    if (subMod.getClass().isAnnotationPresent(Mod.EventBusSubscriber.class))
+                    {
+                        eventSubscriberWarns.add(container);
+                        MinecraftForge.EVENT_BUS.unregister(subMod.getClass());
+                    }
+
+                    if (CommunityMod.getConfig().isSubModEnabled(container))
+                    {
+                        loadedSubMods.add(container);
+                        MinecraftForge.EVENT_BUS.register(subMod.getClass());
+
+                        LOGGER.info("Loaded submod {} from class {}.", modName, data.getClassName());
+                    }
+                    else if (!requiresMcRestart)
```

Comment:
Why does requiring a restart prevent it from being registered into the event bus, these seem like completely isolated properties?

Reply:
Since event handlers can't be unregistered, they're only registered for that submod if
A) it's loaded during startup (L56), or
B) it won't require a restart.

Submods that don't require a restart are expected to limit their features to when they're enabled, so they get their handlers registered on init regardless if they're enabled or not.